File: airbyte-integrations/connectors/destination-gcs/src/test-integration/java/io/airbyte/integrations/destination/gcs/GcsAvroParquetDestinationAcceptanceTest.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.gcs;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.commons.json.Jsons;
import io.airbyte.commons.resources.MoreResources;
import io.airbyte.integrations.destination.s3.S3Format;
import io.airbyte.integrations.destination.s3.avro.JsonSchemaType;
import io.airbyte.integrations.standardtest.destination.ProtocolVersion;
import io.airbyte.integrations.standardtest.destination.argproviders.NumberDataTypeTestArgumentProvider;
import io.airbyte.protocol.models.v0.AirbyteCatalog;
import io.airbyte.protocol.models.v0.AirbyteMessage;
import io.airbyte.protocol.models.v0.AirbyteStream;
import io.airbyte.protocol.models.v0.CatalogHelpers;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import org.apache.avro.Schema;
import org.apache.avro.Schema.Field;
import org.apache.avro.Schema.Type;
import org.apache.avro.generic.GenericData.Record;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ArgumentsSource;

public abstract class GcsAvroParquetDestinationAcceptanceTest extends GcsDestinationAcceptanceTest {

  public GcsAvroParquetDestinationAcceptanceTest(final S3Format s3Format) {
    super(s3Format);
  }

  @Override
  public ProtocolVersion getProtocolVersion() {
    return ProtocolVersion.V1;
  }

  @ParameterizedTest
  @ArgumentsSource(NumberDataTypeTestArgumentProvider.class)
  public void testNumberDataType(final String catalogFileName, final String messagesFileName) throws Exception {
    final AirbyteCatalog catalog = readCatalogFromFile(catalogFileName);
    final List<AirbyteMessage> messages = readMessagesFromFile(messagesFileName);

    final JsonNode config = getConfig();
    final String defaultSchema = getDefaultSchema(config);
    final ConfiguredAirbyteCatalog configuredCatalog = CatalogHelpers.toDefaultConfiguredCatalog(catalog);
    runSyncAndVerifyStateOutput(config, messages, configuredCatalog, false);

    for (final AirbyteStream stream : catalog.getStreams()) {
      final String streamName = stream.getName();
      final String schema = stream.getNamespace() != null ? stream.getNamespace() : defaultSchema;

      final Map<String, Set<Type>> actualSchemaTypes = retrieveDataTypesFromPersistedFiles(streamName, schema);
      final Map<String, Set<Type>> expectedSchemaTypes = retrieveExpectedDataTypes(stream);

      assertEquals(expectedSchemaTypes, actualSchemaTypes);
    }
  }

  private Map<String, Set<Type>> retrieveExpectedDataTypes(final AirbyteStream stream) {
    final Iterable<String> iterableNames = () -> stream.getJsonSchema().get("properties").fieldNames();
    final Map<String, JsonNode> nameToNode = StreamSupport.stream(iterableNames.spliterator(), false)
        .collect(Collectors.toMap(
            Function.identity(),
            name -> getJsonNode(stream, name)));

    return nameToNode
        .entrySet()
        .stream()
        .collect(Collectors.toMap(
            Entry::getKey,
            entry -> getExpectedSchemaType(entry.getValue())));
  }

  private JsonNode getJsonNode(final AirbyteStream stream, final String name) {
    final JsonNode properties = stream.getJsonSchema().get("properties");
    if (properties.size() == 1) {
      return properties.get("data");
    }
    return properties.get(name).get("items");
  }

  private Set<Type> getExpectedSchemaType(final JsonNode fieldDefinition) {
    // The $ref is a migration to V1 data type protocol see well_known_types.yaml
    final JsonNode typeProperty = fieldDefinition.get("type") == null ? fieldDefinition.get("$ref") : fieldDefinition.get("type");
    final JsonNode airbyteTypeProperty = fieldDefinition.get("airbyte_type");
    final String airbyteTypePropertyText = airbyteTypeProperty == null ? null : airbyteTypeProperty.asText();
    return Arrays.stream(JsonSchemaType.values())
        .filter(
            value -> value.getJsonSchemaType().equals(typeProperty.asText()) && compareAirbyteTypes(airbyteTypePropertyText, value))
        .map(JsonSchemaType::getAvroType)
        .collect(Collectors.toSet());
  }

  private boolean compareAirbyteTypes(final String airbyteTypePropertyText, final JsonSchemaType value) {
    if (airbyteTypePropertyText == null) {
      return value.getJsonSchemaAirbyteType() == null;
    }
    return airbyteTypePropertyText.equals(value.getJsonSchemaAirbyteType());
  }

  private AirbyteCatalog readCatalogFromFile(final String catalogFilename) throws IOException {
    return Jsons.deserialize(MoreResources.readResource(catalogFilename), AirbyteCatalog.class);
  }

  private List<AirbyteMessage> readMessagesFromFile(final String messagesFilename) throws IOException {
    return MoreResources.readResource(messagesFilename).lines()
        .map(record -> Jsons.deserialize(record, AirbyteMessage.class)).collect(Collectors.toList());
  }

  protected abstract Map<String, Set<Type>> retrieveDataTypesFromPersistedFiles(final String streamName, final String namespace) throws Exception;

  protected Map<String, Set<Type>> getTypes(final Record record) {

    final List<Field> fieldList = record
        .getSchema()
        .getFields()
        .stream()
        .filter(field -> !field.name().startsWith("_airbyte"))
        .toList();

    if (fieldList.size() == 1) {
      return fieldList
          .stream()
          .collect(
              Collectors.toMap(
                  Field::name,
                  field -> field.schema().getTypes().stream().map(Schema::getType).filter(type -> !type.equals(Type.NULL))
                      .collect(Collectors.toSet())));
    } else {
      return fieldList
          .stream()
          .collect(
              Collectors.toMap(
                  Field::name,
                  field -> field.schema().getTypes()
                      .stream().filter(type -> !type.getType().equals(Type.NULL))
                      .flatMap(type -> type.getElementType().getTypes().stream()).map(Schema::getType).filter(type -> !type.equals(Type.NULL))
                      .collect(Collectors.toSet())));
    }
  }

}

File: airbyte-cdk/java/airbyte-cdk/gcs-destinations/src/testFixtures/kotlin/io/airbyte/cdk/integrations/destination/gcs/GcsAvroParquetDestinationAcceptanceTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.gcs

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.cdk.integrations.destination.s3.FileUploadFormat
import io.airbyte.cdk.integrations.destination.s3.avro.JsonSchemaType
import io.airbyte.cdk.integrations.standardtest.destination.ProtocolVersion
import io.airbyte.cdk.integrations.standardtest.destination.argproviders.NumberDataTypeTestArgumentProvider
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.resources.MoreResources
import io.airbyte.protocol.models.v0.AirbyteCatalog
import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.AirbyteStream
import io.airbyte.protocol.models.v0.CatalogHelpers
import java.io.IOException
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.StreamSupport
import org.apache.avro.Schema
import org.apache.avro.generic.GenericData
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ArgumentsSource

abstract class GcsAvroParquetDestinationAcceptanceTest(fileUploadFormat: FileUploadFormat) :
    GcsDestinationAcceptanceTest(fileUploadFormat) {
    override fun getProtocolVersion() = ProtocolVersion.V1

    @ParameterizedTest
    @ArgumentsSource(NumberDataTypeTestArgumentProvider::class)
    @Throws(Exception::class)
    fun testNumberDataType(catalogFileName: String, messagesFileName: String) {
        val catalog = readCatalogFromFile(catalogFileName)
        val messages = readMessagesFromFile(messagesFileName)

        val config = this.getConfig()
        val defaultSchema = getDefaultSchema(config)
        val configuredCatalog = CatalogHelpers.toDefaultConfiguredCatalog(catalog)
        runSyncAndVerifyStateOutput(config, messages, configuredCatalog, false)

        for (stream in catalog.streams) {
            val streamName = stream.name
            val schema = if (stream.namespace != null) stream.namespace else defaultSchema!!

            val actualSchemaTypes = retrieveDataTypesFromPersistedFiles(streamName, schema)
            val expectedSchemaTypes = retrieveExpectedDataTypes(stream)

            Assertions.assertEquals(expectedSchemaTypes, actualSchemaTypes)
        }
    }

    private fun retrieveExpectedDataTypes(stream: AirbyteStream): Map<String, Set<Schema.Type>> {
        val iterableNames = Iterable { stream.jsonSchema["properties"].fieldNames() }
        val nameToNode =
            StreamSupport.stream(iterableNames.spliterator(), false)
                .collect(
                    Collectors.toMap(
                        Function.identity(),
                        Function { name: String -> getJsonNode(stream, name) }
                    )
                )

        return nameToNode.entries.associate { it.key to getExpectedSchemaType(it.value) }
    }

    private fun getJsonNode(stream: AirbyteStream, name: String): JsonNode {
        val properties = stream.jsonSchema["properties"]
        if (properties.size() == 1) {
            return properties["data"]
        }
        return properties[name]["items"]
    }

    private fun getExpectedSchemaType(fieldDefinition: JsonNode): Set<Schema.Type> {
        // The $ref is a migration to V1 data type protocol see well_known_types.yaml
        val typeProperty =
            if (fieldDefinition["type"] == null) fieldDefinition["\$ref"]
            else fieldDefinition["type"]
        val airbyteTypeProperty = fieldDefinition["airbyte_type"]
        val airbyteTypePropertyText = airbyteTypeProperty?.asText()
        return JsonSchemaType.entries
            .toTypedArray()
            .filter { value: JsonSchemaType ->
                value.jsonSchemaType == typeProperty.asText() &&
                    compareAirbyteTypes(airbyteTypePropertyText, value)
            }
            .map { it.avroType }
            .toSet()
    }

    private fun compareAirbyteTypes(
        airbyteTypePropertyText: String?,
        value: JsonSchemaType
    ): Boolean {
        if (airbyteTypePropertyText == null) {
            return value.jsonSchemaAirbyteType == null
        }
        return airbyteTypePropertyText == value.jsonSchemaAirbyteType
    }

    @Throws(IOException::class)
    private fun readCatalogFromFile(catalogFilename: String): AirbyteCatalog {
        return Jsons.deserialize(
            MoreResources.readResource(catalogFilename),
            AirbyteCatalog::class.java
        )
    }

    @Throws(IOException::class)
    private fun readMessagesFromFile(messagesFilename: String): List<AirbyteMessage> {
        return MoreResources.readResource(messagesFilename).lines().map {
            Jsons.deserialize(it, AirbyteMessage::class.java)
        }
    }

    @Throws(Exception::class)
    protected abstract fun retrieveDataTypesFromPersistedFiles(
        streamName: String,
        namespace: String
    ): Map<String?, Set<Schema.Type?>?>

    protected fun getTypes(record: GenericData.Record): Map<String, Set<Schema.Type>> {
        val fieldList =
            record.schema.fields.filter { field: Schema.Field ->
                !field.name().startsWith("_airbyte")
            }

        return if (fieldList.size == 1) {
            fieldList.associate {
                it.name() to
                    it.schema()
                        .types
                        .map { obj: Schema -> obj.type }
                        .filter { type: Schema.Type -> type != Schema.Type.NULL }
                        .toSet()
            }
        } else {
            fieldList.associate {
                it.name() to
                    it.schema()
                        .types
                        .filter { type: Schema -> type.type != Schema.Type.NULL }
                        .flatMap { type: Schema -> type.elementType.types }
                        .map { obj: Schema -> obj.type }
                        .filter { type: Schema.Type -> type != Schema.Type.NULL }
                        .toSet()
            }
        }
    }
}
